module/VuFind/src/VuFind/ILS/Driver/SierraRest.php

```diff
@@ -914,32 +935,70 @@ public function updateHolds(
     ): array {
         $results = [];
         foreach ($holdsDetails as $requestId) {
+            // Fetch existing hold status:
+            $reqFields = 'status,pickupLocation,frozen'
+                . (isset($fields['frozen']) ? ',canFreeze' : '');
+            $hold = $this->makeRequest(
+                [$this->apiBase, 'patrons', 'holds', $requestId],
+                [
+                    'fields' => $reqFields
+                ],
+                'GET',
+                $patron
+            );
+            $available = in_array($hold['status']['code'], ['b', 'j', 'i']);
+            $inTransit = $hold['status']['code'] === 't';
+
+            // Check if we can do the requested changes:
             $updateFields = [];
-            if (isset($fields['frozen'])) {
-                $updateFields['freeze'] = $fields['frozen'];
+            $fieldsSkipped = false;
+            if (isset($fields['frozen']) && $hold['frozen'] !== $fields['frozen']) {
+                if ($fields['frozen'] && !$hold['canFreeze']) {
+                    $fieldsSkipped = true;
+                } else {
+                    $updateFields['freeze'] = $fields['frozen'];
+                }
             }
             if (isset($fields['pickUpLocation'])) {
-                $updateFields['pickupLocation'] = $fields['pickUpLocation'];
+                if ($available || $inTransit) {
+                    $fieldsSkipped = true;
+                } else {
+                    $updateFields['pickupLocation'] = $fields['pickUpLocation'];
+                }
             }
 
-            $result = $this->makeRequest(
-                [$this->apiBase, 'patrons', 'holds', $requestId],
-                json_encode($updateFields),
-                'PUT',
-                $patron
-            );
-
-            if (!empty($result['code'])) {
+            if (!$updateFields) {
                 $results[$requestId] = [
                     'success' => false,
-                    'status' => $this->formatErrorMessage(
-                        $result['description'] ?? $result['name']
-                    )
+                    'status' => 'hold_error_update_blocked_status'
                 ];
             } else {
-                $results[$requestId] = [
-                    'success' => true
-                ];
+                $result = $this->makeRequest(
+                    [$this->apiBase, 'patrons', 'holds', $requestId],
+                    json_encode($updateFields),
+                    'PUT',
+                    $patron
+                );
+
+                if (!empty($result['code'])) {
+                    $results[$requestId] = [
+                        'success' => false,
+                        'status' => $this->formatErrorMessage(
+                            $result['description'] ?? $result['name']
+                        )
+                    ];
+                } else {
```

Comment:
Is there a compelling reason to nest here -- would this be a little more readable as a single `if .. elseif .. else` clause?

Reply:
No reason, fixed!